src/scripts/ganeshactl/ganesha_conf.man
.\" This is a comment
.\" This is a comment
.TH ganesha_conf 8 "21 Mar 2017"
.SH NAME
ganesha_conf \- Ganesha configuration editor

.SH SYNOPSIS
.B ganesha_conf set
.I <block-description>
.B [--key1 value1 [--key2 value2]...]
.br
.B ganesha_conf del
.I <block-description>
.B [--key1 [--key2 ]...]

.SH DESCRIPTION
.PP
The
.B set
command adds or changes the given key value parameters in
the block. It also creates the block if the block corresponding to
the given
.I block description
(see section
.B BLOCK DESCRIPTION
below)
is not present in the configuration.
.PP
The
.B del
command deletes given keys from the ganesha configuration
block described by the block description. It will delete the block
itself if no keys are provided.

.SH BLOCK DESCRIPTION:
.PP
.I Block description
is a list of block names and possibly some key value parameters that
uniquely identify a ganesha configuration block.
.PP
NFS Ganesha configuration file contains a list of blocks. Each
block starts with a block
.B name
followed by a
.B left brace,
then a list of
.B key = value;
entries. The block may optionally have
.B sub blocks
(note the recursive definition!).  Finally, the block
ends with a
.B right brace.
(Note that
.B key = value;
entries can come after a sub block as well, but we don't allow this with
ganesha_conf editor! All key value entries should come before any sub
blocks.)

An example of a ganesha configuration block:

.PP
.in +4n
.nf
nfs_core_param {
    Nb_Worker = 256;
    Clustered = TRUE;
    NFS_Protocols = 3,4;
}
.fi
.in
.PP
Since there should be only one
.B nfs_core_param
block, we just need the name of the block to uniquely identify it. So
"nfs_core_param" would be its block description!
.PP
An example of a ganesha configuration block with a couple sub blocks:
.PP
.in +4n
.nf
log {
    default_log_level = EVENT;
    format {
        date_format = ISO-8601;
        time_format = ISO-8601;
        thread_name = TRUE;
    }
    components {
        all = EVENT;
    }
}
.fi
.in
.PP
Ganesha configuration should have only one
.B log
block as well, so "log" would identify the log block. To identify
.B format
sub block inside the
.B log
block, "log format" would be the block description for the above
.B format
sub block. Similarly "log components" would be the block description for
the above
.B components
sub block.
.PP
An
.B export
block is special in that there can be many export
blocks, one for each export. A
.B client
block is also special. There can be many
.B client
blocks and they are
always sub blocks inside
.B export
blocks.
.I export_id
key value uniquely identifies an
.B export
block.
.I clients
key value uniquely identifies a
.B client
block inside a given
.B export
block.
.PP
Here are couple
.B export
blocks with couple
.B client
blocks in them:
.PP
.in +4n
.nf
export {
    export_id = 1;
    path = /fs1/export1;
    pseudo = /fs1/export1;
    manage_gids = true;
    client {
        clients = 192.168.122.31;
        access_type = RW;
    }
    client {
        clients = *;
        access_type = RO;
    }
}

export {
    export_id = 2;
    path = /fs1/export2;
    pseudo = /fs1/export2;
    manage_gids = true;
    client {
        clients = 192.168.122.31;
        access_type = RW;
    }
    client {
        clients = 192.168.122.32;
        access_type = RO;
    }
}
.fi
.in
.PP
To identify the correct
.B export
block, we need to supply its
.I export_id.
For example "export export_id 2" identifies the second
.B export
block above.
.B export
blocks can be uniquely identified by
.I pseudo
or
.I path
keys in some environments. One could specify
"export path /fs1/export2" to identify the second
.B export
block as well.  Similarly, a
.B client
block needs additional
.I clients
key value to identify the correct
.B client
block. For example, "export export_id 2 client clients 192.168.122.31" identifies the first
.B client
block in the second
.B export
block above!

.SH EXAMPLES:
1. To change number of ganesha worker threads:
.PP
ganesha_conf set nfs_core_param --nb_worker 256
.PP
2. To change the date and time format of ganesha log messages:
.PP
ganesha_conf set log format --date_format ISO-8601 --time_format ISO-8601
.PP
3. Create an export and allow client with IP address 192.168.122.31
to be able to do read write and all other clients to do read only:
.PP
ganesha_conf set export path /fs1/export2 --export_id 2 --pseudo /fs1/export2 --manage_gids true
.br
ganesha_conf set export path /fs1/export2 client clients 192.168.122.31 --access_type RW
.br
ganesha_conf set export path /fs1/export2 client clients 192.168.122.32 --access_type RO

.SH NOTES:
.B ganesha_conf
by default uses /etc/ganesha/ganesha.conf file as the
configuration file. If your environment uses a different file (or
set of files), you can use
.B CONFFILE
environment variable to override the default configuration file.
For example,
"CONFFILE=/etc/ganesha/ganesha.main.conf ganesha_conf set nfs_core_param --nb_worker 256"
will use /etc/ganesha/ganesha.main.conf file for changing the worker threads.
.PP
.B ganesha_conf
can't handle comments within a block at this point.
.PP
Neither block descriptions nor key value parameters are verified to
be valid ganesha configuration blocks or parameter values currently.


src/ChangeLog
0.99.39 - Mon 21 Jul 2008
	- Initial version submitted to sourceforge

0.99.40 - Mon 28 Jul 2008 
	- Fixed several memory issues in NFSv4 implementation 
	- Fixed several memory issues in FSAL_PROXY
	- Fixed a badly placed Mem_Free in support/nfs_state_id.c (function nfs4_State_Del)
        - Added Handle Mapping feature in FSAL_PROXY which makes it possible to export
          back in NFSv2 and NFSv3 from a proxyfied server accessed via NFSv4

0.99.41 - Mon 18 Aug 2008
	- Fixed nfs4_op_access bug due to bad interpretation of the RFC for
	  OP_ACCESS4
	- Added extended features in BuddyMalloc module to enable extended
	  memory leak tracking (run the ./configure script with the following
	  options : "--enable-debug-memleaks --disable-block-prealloc")
	- Fixed a bug in FSAL_PROXY that made every user have root permissions
	  in a few situations
	- Bug fixed: bad offset management in FSAL_read/FSAL_write for
	  FSAL_PROXY. This made the datacache behaves weirdly. 
	- Use All-0 stateid for r/w operations made for maintaining the data
	  cache coherent with the "proxyfied" server with FSAL_PROXY. Due to
	  this the parameter NFSv4_Proxy::Open_by_FH_Working_Dir is no more
	  required, making the configuration of nfs-ganesha as a proxy much
	  easier
	- FSAL_PROXY now supports RPCSEC_GSS authentication

0.99.42 - Mon 29 Sep 2008
	- Added xattr support in NFSv4. For object "foo" a ghost
	  directory name ".xattr.d.foo" is used to access extended attributes
        - Added xattr ghost directory and ghost objects for NFSv3. These "extended
	  attributes" are read-only for the moment. These two new features may
          be disable if "--disable-xattr-directory" is used at "./configure" time
        - Enhancement in FSAL_GetXattrs function (file FSAL/FSAL_*/fsal_xattrs.c) for
          making integration to xattr.d easier

0.99.43 - Fri 10 Oct 2008
        - Fixed RPCSEC_GSS support. It is now possible to mount with krb5
	  authentication. Security tuple krb5, krb5i and krb5p are supported
	- Add RPCSEC_GSS specifc feature for NFSv4
	- Add RPCSEC_GSS specifc feature for NFSv3
	- Configuration file supports "nodesets" syntax( the set
     	  "node1,node3,node4,node5,node6" can be summarized as
	  "node[1,3,4-6]). This feature is very useful if your client is a
	  cluster.
	- New RPM packaging. 
	- The libganeshaNFS that provide NFS exports for FUSE ready product is
	  now available as both static and shared libraries. 

0.99.44 - Mon 20 Oct 2008 
        - NFS4_OP_OPEN default behaviour is now to ask the client to confirm 
	  the open using NFS5_OP_OPEN_CONFIRM. The default behaviour was
	  formerly the opposite.
	- A bug in NFS4_OP_OPEN : same open_owner could have multiple states
	  on the same file in certain cases 
	- Many bug fixed in NFS4_OP_LOCK
	- libganeshaNFS (produced when FSAL_FUSELIKE is chosen) can now be
	  managed by pkgconfig (a ".pc" file is added to the install target)
	- Improved compatibility for old FUSE filesystems that implement getdir
	  instead of readdir.
	- FUSE-like binding now support filesystems with no inode numbers.
	- TIRPC can be use as alternate RPC layer. In later releases, this
	  feature will be used to support IPv6 with NFS-GANESHA.
	- Added a 'init' scripts to be place in /etc/init.d 
	- Added stuff to specfile so that it installs the .vim files

0.99.45 - Wed  5 Nov 2008 
	- Add IPv6 support via TIRPC support 
	- Add features to configuration file parsing for IPv6 address
	- Bug fix in handle mapping module for FSAL_PROXY
	- Bug Fix: Bad verifier management in FSAL_PROXY's cleintid
	  negociation
	- Bug fix: bad management of uid=0 is uidgidmap cache when using
	  RPCSEC_GSS authentication.
	- Bug fix: several bug fixed in .xattr.d management for NFSv3

0.99.46 - Wed 18 Nov 2008 	
	- Bug fix: possible deadlock may occur in worker thread when using tcp
	  clients
	- Minor feature: MD cache now uses home made RW_Lock (the same that 
	  Hashtable module already uses). Call rw_lock_downgrade was added to
	  RW_Lock module
	- DataCache GC is done by an external call to the ganesha service
	  started in "syncers" of "flushers" mode.

0.99.47 - Thu  4 Dec 2008
	- Running the Spec benchmark on nfs-ganesha showed a bug in attributes
	  asked at creation time in NFSPROC3_CREATE, this has been fixed
	- Fix non-compliancy (thanks again to Spec NFS) in NFSPROC3_SYMLINK (attributes provided as 
	  arguments were never used).
	- Code cleanup: removed structure related to deprecated way of
	  performing datacache's garbagge collection.
	- Major feature added : skeleton and "basements" added for MFSL, a
	  new module that goes between md cache and FSAL 
	- Removed direct Cache_Inode_Async, md writeback will be provided in a
	  future module MFSL_ASYNC in MFSL directory 
	 
0.99.48	- Mon 15 Dec 2008
	- Renamed yyparse and yylex functions so that GANESHA/FUSE would no
	  more interfer with FUSE modules that uses lex/yacc parsing
	- FSAL_PROXY was ported to MacOS X (Darwin 9.5.0)
	- FSAL_FUSE was ported to MacOS X (Darwin 9.5.0). This allow user
	  space libs with fuse binding to be used from MacOS through
	  NFS-GANESHA
	- FSAL_SNMP has been ported to MacOS X (Darwin 9.5.0)
	- FSAL_POSIX has been ported to MacOS X (Darwin 9.5.0)
	 
0.99.49	- Fri  9 Jan 2009  
        - FSAL_PROXY has been ported to FreeBSD 7.0
        - FSAL_FUSE has been ported to FreeBSD 7.0 (FUSE ready product can now
	  export in NFS via NFS-GANESHA)
        - FSAL_POSIX has been ported to FreeBSD 7.0
        - FSAL_SNMP has been ported to FreeBSD 7.0
	- Code has been ported and validated to ia64 architecture for all
	  FSALs

0.99.50 - Fri 23 Jan 2009
	- Code has been validated on PPC/Linux architecture 
	- Fixed a Makefile.am that prevent from compiling with FSAL_FUSE and
	  gssrpc activated
	- Added a 'debian' directory in order to produce debian packages
	- FSAL_POSIX has been ported to OpenSolaris (alpha)
	- FSAL_FUSE has been ported to OpenSolaris (beta)

src/scripts/ganeshactl/ganesha_conf.py
#!/usr/bin/python3
# SPDX-License-Identifier: LGPL-3.0-or-later
#
# Copyright (C) 2017 IBM
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Author: Malahal Naineni <malahal@us.ibm.com>

import os, sys
from Ganesha.config_editor import BLOCK, ArgError
import logging, pprint

# Modify the file with the given data atomically
def modify_file(filename, data):
    from tempfile import NamedTemporaryFile
    tmp = NamedTemporaryFile(dir=os.path.dirname(filename), delete=False)
    tmp.write(data)
    tmp.flush()
    os.fsync(tmp.fileno())

    # If filename exists, get its stats and apply them to the temp file
    try:
        stat = os.stat(filename)
        os.chown(tmp.name, stat.st_uid, stat.st_gid)
        os.chmod(tmp.name, stat.st_mode)
    except:
        pass

    os.rename(tmp.name, filename)

# Get block names and key value options as separate lists
def get_blocks(args):
    key_found = False
    for i, arg in enumerate(args):
        if arg.startswith("--"):
            key_found = True
            break

    if key_found:
        return (args[0:i], args[i:])

    return (args, [])


usage = """
Usage: %s set <block-descriptor> [--param value]
       %s del <block-descriptor> [--param]

       where <block-descriptor> is a list of blocknames with possible key value pair identifying the block.

       For example,
       %s set log --default_log_level DEBUG
       %s set log components --FSAL FULL_DEBUG
       %s set export export_id 14 --pseudo "/nfsroot/export1"
       %s set export export_id 14 client clients '*' --manage-gids true

       %s get cacheinode
       %s get cacheinode --Dir_Max
""" % (8 * (sys.argv[0],))

if len(sys.argv) < 2:
    sys.exit(usage)

FORMAT = "[%(filename)s:%(lineno)d-%(funcName)s()] %(message)s"
#logging.basicConfig(format=FORMAT, level=logging.DEBUG)

opcode = sys.argv[1]
if opcode == "set":
    (names, pairs) = get_blocks(sys.argv[2:])
    if not names:
        sys.exit("no block names")
    if len(pairs) % 2 != 0:
        sys.exit("odd number for key value pairs")

    for key in pairs[::2]:
        if not key.startswith("--"):
            sys.exit("some keys are not with -- prefix")

    # remove "--" prefix of keys
    keys = [key[2:] for key in pairs[::2]]
    values = [value for value in pairs[1::2]]

    opairs = zip(keys, values)
    block = BLOCK(names)

elif opcode == "get":
    (names, pairs) = get_blocks(sys.argv[2:])
    if not names:
        sys.exit("no block names")
    if len(pairs) > 1:
        sys.exit("only zero or one value for get operations")
    for key in pairs[::2]:
        if not key.startswith("--"):
            sys.exit("some keys are not with -- prefix")

    # remove "--" prefix of keys
    keys = [key[2:] for key in pairs[::2]]
    block = BLOCK(names)

elif opcode == "del":
    (names, keys) = get_blocks(sys.argv[2:])
    if not names:
        sys.exit("no block names")
    for key in keys:
        if not key.startswith("--"):
            sys.exit("some key not with -- prefix")

    # remove "--" prefix of keys
    keys = [key[2:] for key in keys]
    if not keys:
        keys = []

    block = BLOCK(names)
else:
    sys.exit(usage)

conffile = os.environ.get("CONFFILE", "/etc/ganesha/ganesha.conf")
old = open(conffile).read()

try:
    if opcode == "set":
        logging.debug("opairs: %s", pprint.pformat(opairs))
        new = block.set_keys(old, opairs)
    elif opcode == "get":
        logging.debug("keys: %s", pprint.pformat(keys))
        value = block.get_keys(old, keys)
        sys.exit(value)
    else:
        logging.debug("keys: %s", pprint.pformat(keys))
        new = block.del_keys(old, keys)
except ArgError as e:
    sys.exit(e.error)

modify_file(conffile, new)
